File: desktop/timing.py
"""Timing modes. Native is the no-delay default; human pacing is explicit;
compatibility uses one fixed
measured interval for targets that drop fast events.

Wiki research part (2026-09-04): trailing sleeps after the last event
cannot aid delivery, so gaps apply BETWEEN events only — n chars get
n-1 gaps, and movement duration 0 means post all points immediately.
"""

import math
import random
from typing import Iterator, List

MODES = ("human", "compatibility", "native")

TYPE_GAPS = (
    ((0.0, 0.1), 0.377),
    ((0.1, 0.2), 0.5492),
    ((0.2, 0.4), 0.0738),
)

# compatibility interval: measured minimum for event-dropping targets;
# tuned per target later with evidence (wiki open question 3)
COMPAT_GAP_S = 0.01


class TimingMode:
    COMPAT_GAP_S = COMPAT_GAP_S


def _validate(mode: str) -> str:
    if mode not in MODES:
        raise ValueError(f"unknown timing mode: {mode} (use {', '.join(MODES)})")
    return mode


def next_type_gap(rng: random.Random) -> float:
    draw = rng.uniform(0, 1)
    cumulative = 0.0
    for (low, high), probability in TYPE_GAPS:
        cumulative += probability
        if draw <= cumulative:
            return rng.uniform(low, high)
    low, high = TYPE_GAPS[-1][0]
    return rng.uniform(low, high)


def gaps_for(mode: str, text: str) -> List[float]:
    """One gap per character. Trailing gap after the last char is dropped
    by the caller; native yields all zeros."""
    _validate(mode)
    if mode == "native":
        return [0.0] * len(text)
    if mode == "compatibility":
        return [COMPAT_GAP_S] * len(text)
    rng = random.Random()  # human mode is explicitly not deterministic
    return [next_type_gap(rng) for _ in text]


def movement_seconds(mode: str, x0: int, y0: int, x1: int, y1: int,
                     target_size: float = 80.0) -> float:
    """Total move duration in seconds. Native = 0 (post immediately).
    Negative results clamp to 0.0."""
    _validate(mode)
    if mode == "native":
        return 0.0
    from .move import duration

    d = duration(x0, y0, x1, y1, target_size)
    return max(0.0, d)

File: desktop/rust/src/timing.rs
use crate::motion::{PythonRandom, movement_time};
use std::time::{SystemTime, UNIX_EPOCH};

pub const MODES: &[&str] = &["human", "compatibility", "native"];
const GAPS: [((f64, f64), f64); 3] = [
    ((0.0, 0.1), 0.377),
    ((0.1, 0.2), 0.5492),
    ((0.2, 0.4), 0.0738),
];

pub fn gaps_for(mode: &str, text: &str) -> Result<Vec<f64>, String> {
    let seed = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_nanos() as i64;
    gaps_for_seed(mode, text, seed)
}

/// Seeded form for deterministic parity tests; normal human timing uses a fresh clock seed.
pub fn gaps_for_seed(mode: &str, text: &str, seed: i64) -> Result<Vec<f64>, String> {
    match mode {
        "native" => Ok(vec![0.0; text.chars().count()]),
        "compatibility" => Ok(vec![0.01; text.chars().count()]),
        "human" => {
            let mut rng = PythonRandom::new(seed);
            Ok(text.chars().map(|_| next_type_gap(&mut rng)).collect())
        }
        _ => Err(format!(
            "unknown timing mode: {mode} (use human, compatibility, native)"
        )),
    }
}

pub fn movement_seconds(
    mode: &str,
    start: (i64, i64),
    end: (i64, i64),
    target_size: f64,
) -> Result<f64, String> {
    if !MODES.contains(&mode) {
        return Err(format!(
            "unknown timing mode: {mode} (use human, compatibility, native)"
        ));
    }
    if mode == "native" {
        return Ok(0.0);
    }
    let distance = ((end.0 - start.0) as f64).hypot((end.1 - start.1) as f64);
    Ok(movement_time(distance, target_size).max(0.0))
}

fn next_type_gap(rng: &mut PythonRandom) -> f64 {
    let draw = rng.random();
    let mut cumulative = 0.0;
    for ((low, high), probability) in GAPS {
        cumulative += probability;
        if draw <= cumulative {
            return low + (high - low) * rng.random();
        }
    }
    let (low, high) = GAPS[2].0;
    low + (high - low) * rng.random()
}
